Started drafting the skeleton of the request management filter

MikeSpreitzer opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test

/kind feature

> /kind flake

**What this PR does / why we need it**:
This PR introduces the top level of the new filter.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
This PR does not fix but contributes to https://github.com/kubernetes/enhancements/issues/1040

**Special notes for your reviewer**:
This PR builds on #77048 

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
The full feature will definitely require a release note.
```

MikeSpreitzer committed:
Started drafting skeleton of the request management filter

MikeSpreitzer commented:
/uncc @caesarxuchao 
/uncc @brendanburns 
/cc @yue9944882 
/cc @lavalamp 
/cc @deads2k 
/cc @aaron-prindle 
/cc @wojtek-t 
/cc @wu8685 
/cc @adohe 
/cc @KevinWang15

fedebongio commented:
/assign @aaron-prindle

aaron-prindle commented:
for reference: https://github.com/kubernetes/kubernetes/issues/76846

MikeSpreitzer commented:
@aaron-prindle: yes, I know that the skeleton work is officially on your plate, and would be happy to see you take it from here.  I see that you have been focused on doing some bottom-up work, and this PR actually originated as an attempt at a particularly constructive review of that.  I wanted to work out my opinion of how we should organize our queuing and dispatching concepts to fit into the context of making an apiserver filter.  That is the spirit in which this is offered.

aaron-prindle commented:
related feature: https://github.com/kubernetes/kubernetes/issues/76846

aaron-prindle commented:
@MikeSpreitzer this is great!  My original comment was just for keeping tabs on all the work being done in the Umbrella Tracking issue.

MikeSpreitzer commented:
Tracked recent work on API object types.

deads2k commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -0,0 +1,242 @@
+
+var reqMgmt *requestManagement
+
+var rmStartOnce sync.Once
```
We don't have globals in kube.  We have instances attached to the structs we use to launch.

MikeSpreitzer replied:
fixed.

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -0,0 +1,242 @@
+	return tr.Kind + "/" + tr.Name
+}
+
+var reqMgmt *requestManagement
```
singleton

MikeSpreitzer replied:
fixed

deads2k commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -0,0 +1,242 @@
+
+// rmSetup is invoked once at startup to put a useful value in `reqMgmt`
+func rmSetup(loopbackClientConfig *restclient.Config, serverConcurrencyLimit int, requestWaitLimit time.Duration) {
+	reqMgmt = &requestManagement{
```
wiring a global to state for a particular instance is something we avoid.

MikeSpreitzer replied:
fixed

MikeSpreitzer commented:
I removed the global variables and updated the bazel.

aaron-prindle commented:
Not sure if it makes sense for this PR but I think that the skeleton might require some additional logic to handle the context from the timeout handler.

MikeSpreitzer commented:
Yes, handling the timeout from the upstream timeout filter _does_ belong here.  The wait duration limit in the queue should be set to something well under the upstream request timeout, so we do not have to worry about that timeout expsiring while the request is waiting in the queue.  The thing we _should_ do here is at the point where the skeleton invokes the remainder of the handler chain, it should use goroutines and `select` to wait for the soonest of either (a) completion of the regular handling of the request or (b) expiration of the upstream timeout.

So, to be precise, what I am thinking is that instead of https://github.com/MikeSpreitzer/kubernetes/blob/8bb3d2ea21d798a96cade4382a02374569aae29f/staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go#L219 where it says
```
			handler.ServeHTTP(w, r)
```
we should instead do this:
```
			timedOut := ctx.Done()
			finished := make(chan struct{})
			go func() {
				handler.ServeHTTP(w, r)
				close(finished)
			}()
			select {
			case <-timedOut:
				klog.V(5).Infof("Timed out waiting for %v to finish\n", r)
			case <-finished:
			}
```

MikeSpreitzer commented:
/retest
The failure must be a flake, this PR only introduces code that is not used.

MikeSpreitzer commented:
I updated the skeleton to handle timeouts from upstream, and squashed.

MikeSpreitzer commented:
/retest

MikeSpreitzer commented:
/retest

MikeSpreitzer commented:
/retest

MikeSpreitzer committed:
update bazel by hack

MikeSpreitzer commented:
/test pull-kubernetes-integration

MikeSpreitzer commented:
My last push was rebasing after the API object types PR (#77048) merged.

MikeSpreitzer committed:
Expanded comment on FairQueuingSystem::DoOnEmpty

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -0,0 +1,264 @@
+	handler http.Handler,
+	loopbackClientConfig *restclient.Config,
+	serverConcurrencyLimit int,
+	requestWaitLimit time.Duration,
```
`requestWaitLimit` is this meant for a time-out for those requests pending in queues? there's already a time-out handler filter, we probably don't want them to overlap..

MikeSpreitzer replied:
This parameter is introduced in https://github.com/kubernetes/enhancements/blob/79119ce549cda9cdfdb532273475ea5090fa5415/keps/sig-api-machinery/20190228-priority-and-fairness.md#latency-protection .  In particular, I think it must be set to something substantially shorter than the limit imposed by the timeout filter.  It does not make sense to allow a request to wait in a queue for longer than the total time allowed for waiting+executing.  In fact, I think the wait duration limit should be short enough that the majority of the overall duration limit is available for request execution.

yue9944882 replied:
>  I think it must be set to something substantially shorter than the limit imposed by the timeout filter

i see, we can check that when validating the flags, what do you think will be the default value for this timeout?

MikeSpreitzer replied:
I suggest scaling it with the upstream timeout.  I suggest setting the wait duration limit to 1/4 of the upstream timeout.

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -0,0 +1,264 @@
+func WithRequestManagement(
+	handler http.Handler,
+	loopbackClientConfig *restclient.Config,
+	serverConcurrencyLimit int,
```
>      --max-mutating-requests-inflight int           The maximum number of mutating requests in flight at a given time. When the server exceeds this, it rejects requests. Zero for no limit. (default 200)
>      --max-requests-inflight int                    The maximum number of non-mutating requests in flight at a given time. When the server exceeds this, it rejects requests. Zero for no limit. (default 400)

iiuc, the limit value reads from these flags, amiright? `serverConcurrencyLimit` = `max-mutating-requests-inflight` + `max-requests-inflight`

MikeSpreitzer replied:
See https://github.com/kubernetes/enhancements/blob/79119ce549cda9cdfdb532273475ea5090fa5415/keps/sig-api-machinery/20190228-priority-and-fairness.md#primary-cpu-and-memory-protection , including its last paragraph.

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -0,0 +1,264 @@
+
+// TypedConfigObjectReference is a reference to a relevant config API object.
+// No namespace is needed because none of these objects is namespaced.
+type TypedConfigObjectReference struct {
```
does it have to be a public/exported type?

MikeSpreitzer replied:
I think the division of this stuff among packages is still TBD, so I have not worried about making the right decision for each name yet.

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -0,0 +1,264 @@
+	configQueue workqueue.RateLimitingInterface
+
+	// numConfigWorkers is the configured number of workers for the config queue
+	numConfigWorkers int
```
i think it's a bit overkilling.. it's a good idea to load the config(i.e. PL)  changes in an asynchronous way but do we really need a configurable number of workers in the system? one worker will be sufficient i presume because we don't expect frequent too many PL and PL changes

MikeSpreitzer replied:
Yes, actually I think it is positively good to have just one worker --- that eliminates the need for synchronization among workers.  I will update.

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -0,0 +1,264 @@
+		}
+
+		// Skip tracking long running events.
+		if longRunningRequestCheck != nil && longRunningRequestCheck(r, requestInfo) {
```
we don't need to completely skip long-running requests.. we can start by processing them in a simple way in the beginning: how about we make the long-running requests consume one quota from the PL (and returns it on finishing) but it doesn't has to go through the queues(sharding+fair-queuing). i remember we agree that the point of introducing conconcurrency limit is to address the long-running calls, at least starting from a simple way.

MikeSpreitzer replied:
No, we agreed that these are out of scope.  See the third item in https://github.com/kubernetes/enhancements/blob/79119ce549cda9cdfdb532273475ea5090fa5415/keps/sig-api-machinery/20190228-priority-and-fairness.md#non-goals

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -0,0 +1,264 @@
+				close(finished)
+			}()
+			select {
+			case <-timedOut:
```
let's strip the timeout from the system, another filter will deal with it

MikeSpreitzer replied:
Yes, the upstream timeout filter will return after the timeout expires.  But in this filter we want an upper bound on execution time; my design analyses were predicated on that.  The logic here ensures that there is such an upper bound, as far as this filter is concerned.

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -0,0 +1,264 @@
+	// that entry.  If the priority level becomes desired again before
+	// this is called, the filter cancels by calling `DoOnEmpty(nil)`.
+	// The callback is invoked with no mutexes locked.
+	DoOnEmpty(func())
```
>  If the priority level becomes desired again before
>  this is called, the filter cancels by calling `DoOnEmpty(nil)`.

if a PL gets removed and created again instantly, it should be a fresh new priorityLevelStates instance. twiddling the PL between desired and undesired will make the system complex, isn't it? how do you think about rename this to `Destroy(callback func())` to clarify the usage?

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -0,0 +1,264 @@
+
+	// priorityLevelStates maps the PriorityLevelConfiguration object
+	// name to the state for that level
+	priorityLevelStates map[string]*PriorityLevelState
```
it's better if we use uid as key here? avoiding PL w/ the same name got removed and created instantly?

MikeSpreitzer replied:
Yes, that would be the implementation if we agree on that behavior.  We have not agreed on what the desired behavior is.  https://github.com/kubernetes/enhancements/pull/1098/files#r300193486

MikeSpreitzer reviewed (commented):
In addition to the inline comments: In https://github.com/kubernetes/enhancements/pull/1098 we are debating how to handle delete and re-create of a priority level.  Let's reach agreement on that, then we can make sure the code is consistent with our decision.

MikeSpreitzer committed:
A little more thought about concurrency in the reqmgmt filter

(1) removed the idea of multiple config queue workers; it will be
simpler and adequate if there is only 1.

(2) attempted to clarify the comment on FairQueuingSystem::DoOnEmpty
to (a) make the division of responsibilities clearer, (b) make a
precise statement of the semantics, and (c) be less prescriptive of
how the filter uses this method.

(this is the third edition of this change)

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt-config.go:
```diff
@@ -0,0 +1,284 @@
+		klog.Error("Unable to do initial config sync")
+		return false
+	}
+	go reqMgmt.Run(stopCh)
```
the system should be launched using a post-start-hook and closed using a pre-shutdown-hook. but we can refactor in a follow-up

MikeSpreitzer replied:
Where would I find how to register those hooks?

MikeSpreitzer replied:
I do not see how https://github.com/kubernetes/kubernetes/blob/c7f9dd0bafe2f3328148a85eed9c18322b9f308e/staging/src/k8s.io/apiserver/pkg/server/config.go#L584 can add those hooks, it does not have access to the server.

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt-config.go:
```diff
@@ -0,0 +1,284 @@
+		if oldPriorities[plName] != nil {
+			problem = "undesired"
+		}
+		klog.Warningf("FlowSchema %s references %s priority level %s and will thus not match any requests", fsName, problem, plName)
```
i think we should probably set a condition in the status subresource of the PL in this case? we can do it in a follow-up

MikeSpreitzer replied:
Remember there are multiple apiservers.  I am dubious about each posting status.

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt-config.go:
```diff
@@ -0,0 +1,284 @@
+		switch spec.DistinguisherMethod.Type {
+		case rmtypesv1a1.FlowDistinguisherMethodByUserType, rmtypesv1a1.FlowDistinguisherMethodByNamespaceType:
+		default:
+			klog.Warningf("FlowSchema %s uses unrecognized FlowDistinguisherMethodType %q; the flow distinguisher will be the empty string", fsName, spec.DistinguisherMethod.Type)
```
validation of the model will reject unsupported distinguisher-method, so the "default" case here will never reach

MikeSpreitzer replied:
yes

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt-config.go:
```diff
@@ -0,0 +1,284 @@
+	}
+	ok := reqMgmt.syncOne()
+	if !ok {
+		klog.Error("Unable to do initial config sync")
```
as a convention of golang, https://github.com/golang/go/wiki/Errors#errors, all the error messages should not start w/ a capital letter

MikeSpreitzer replied:
There is a difference between (a) the string in an `error` and (b) a message written with glog or klog.  Yes, the string in an `error` should start lower case, and that is what the cited doc says.  What we have here is (b), where I think a normal English sentence is appropriate.

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt-config.go:
```diff
@@ -0,0 +1,284 @@
+	}
+	reqMgmt.curState.Store(rms)
+	reqMgmt.plInformer.AddEventHandler(reqMgmt)
+	reqMgmt.fsInformer.AddEventHandler(reqMgmt)
```
we should move the event-handler registration below waiting for cache synced, the reason explained here

> if ok := cache.WaitForCacheSync(stopCh, reqMgmt.plInformer.HasSynced, reqMgmt.fsInformer.HasSynced); !ok {

MikeSpreitzer replied:
With the notification handlers fixed the suggested change would make no real difference.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt-config.go:
```diff
@@ -0,0 +1,284 @@
+	}
+}
+
+func applyPriorityLevelSpecDefaults(plName string, spec *rmtypesv1a1.PriorityLevelConfigurationSpec) {
```
yes

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt-config.go:
```diff
@@ -0,0 +1,284 @@
+
+// OnAdd handles notification from an informer of object creation
+func (reqMgmt *requestManagement) OnAdd(obj interface{}) {
+	reqMgmt.configQueue.Add(new(int))
```
`Add(new(int))` was a mistake; I have corrected it to `Add(0)`, meaning that there will be at most one entry in the queue at any given time.

MikeSpreitzer commented:
/retest

MikeSpreitzer committed:
Simplified FairQueuingSystem interface

.. and introduced a tiny bit of starting to work with config objects.

MikeSpreitzer commented:
/retest

yue9944882 commented:
/lgtm

aaron-prindle commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -0,0 +1,316 @@
+			return
+		}
+
+		for {
```
I'm not sure I understand the purpose of the `for` loop here.  Additionally I'm not sure I understand from the skeleton how code exits this loop.

aaron-prindle replied:
I think we might want to 'return' after the select cases

MikeSpreitzer replied:
Oops!  Right!
The purpose of the loop was just to be able to do a retry in the quiescing case.
